README
NAME
    Getopt::Chain - Command-line processing like svn and git

VERSION
    Version 0.017

SYNPOSIS
        package My::Command;

        use Getopt::Chain::Declare;

        start [qw/ verbose|v /]; # These are "global"
                                 # my-command --verbose initialize ...

        # my-command ? initialize ... --> my-command help initialize ...
        rewrite qr/^\?(.*)/ => sub { "help ".($1||'') };

        # NOTE: Rewriting applies to the command sequence, NOT options

        # my-command about ... --> my-command help about
        rewrite [ ['about', 'copying'] ] => sub { "help $1" };

        # my-command initialize --dir=...
        on initialize => [qw/ dir|d=s /], sub {
            my $context = shift;

            my $dir = $context->option( 'dir' )

            # Do initialize stuff with $dir
        };

        # my-command help
        on help => undef, sub {
            my $context = shift;

            # Do help stuff ...
            # First argument is undef because help
            # doesn't take any options
        };

        under help => sub {

            # my-command help create
            # my-command help initialize
            on [ [ qw/create initialize/ ] ] => undef, sub {
                my $context = shift;

                # Do help for create/initialize
                # Both: "help create" and "help initialize" go here
            };

            # my-command help about
            on 'about' => undef, sub {
                my $context = shift;

                # Help for about...
            };

            # my-command help copying
            on 'copying' => undef, sub {
                my $context = shift;

                # Help for copying...
            };

            # my-command help ...
            on qr/^(\S+)$/ => undef, sub {
               my $context = shift;
               my $topic = $1;

               # Catch-all for anything not fitting into the above...

               warn "I don't know about \"$topic\"\n"
            };
        };

        # ... elsewhere ...

        My::Command->new->run( [ @arguments ] )
        My::Command->new->run # Just run with @ARGV

DESCRIPTION
    Getopt::Chain can be used to provide svn(1)- and git(1)-style option and
    command processing. Any option specification covered by Getopt::Long is
    fair game.

    This is a new version of Getopt::Chain that uses Path::Dispatcher

    CAVEAT 1: This is pretty beta, so the sugar/interface above WILL be
    tweaked

    CAVEAT 2: Unfortunately, Getopt::Long slurps up the entire arguments
    array at once. Usually, this isn't a problem (as Getopt::Chain uses
    pass_through). However, if a subcommand has an option with the same name
    or alias as an option for a parent, then that option won't be available
    for the subcommand. For example:

        ./script --verbose --revision 36 edit --revision 48 --file xyzzy.c
        # Getopt::Chain will not associate the second --revision with "edit"

    So, for now, try to use distinct option names/aliases :)

    DEBUG: You can get some extra information about what Getopt::Chain is
    doing by setting the environment variable "GOC_TRACE" to 1

LEGACY
    The old-style, non Path::Dispatcher version is still available at
    Getopt::Chain::v005

SEE ALSO
    Getopt::Long

    App::Cmd

    MooseX::App::Cmd

ACKNOWLEDGEMENTS
    Sartak for Path::Dispatcher

    obra for inspiration on the CLI (via Prophet & Sd:
    <http://syncwith.us/>)

AUTHOR
    Robert Krimen, "<rkrimen at cpan.org>"

SOURCE
    You can contribute or fork this project via GitHub:

    <http://github.com/robertkrimen/Getopt-Chain/tree/master>

        git clone git://github.com/robertkrimen/Getopt-Chain.git Getopt-Chain

BUGS
    Please report any bugs or feature requests to "bug-Getopt-Chain at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Getopt-Chain>. I will be
    notified, and then you'll automatically be notified of progress on your
    bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Getopt::Chain

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Getopt-Chain>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Getopt-Chain>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Getopt-Chain>

    *   Search CPAN

        <http://search.cpan.org/dist/Getopt-Chain>

COPYRIGHT & LICENSE
    Copyright 2008 Robert Krimen, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
TODO:
    - Test with '--' (huh?)
    - Implement 'defer' sugar/backing

0.017 Friday June 12 09:36:24 PDT 2009:
    - Added '--' path conditional and better builder parsing
    - Made ::Context and ::Context::Step more explicit with starting_arguments
      ::Context now uses parsing_arguments (instead of remaining_arguments)
      Reinstituted unparsed option-like arguments as a fatal error (except
      during --)
    
0.016 Friday June 05 09:36:58 PDT 2009:
    - Removed Test::Trap dependency

0.015 Thursday June 04 19:33:25 PDT 2009:
    - Made remainder warning optional on debug (via GOC_TRACE=1)

0.014 Wednesday June 03 17:38:25 PDT 2009:
    - Fixed some documentation
    - Discard unknown, option-like arguments, instead of croaking

0.013_1 Monday June 01 23:14:32 PDT 2009:
    - Allow configuring of the context class
    - Added Declare::under functionality
    - Re-implemented the declarator to use a builder replay
    
0.013 Saturday May 16 00:40:06 PDT 2009:
    - Took erroring XXX out of .t

0.012 Friday May 15 19:45:34 PDT 2009:
    - More saner path handling (* globbing at end)

0.011 Thursday May 14 20:22:44 PDT 2009:
    - Hash::Param dependency in Makefile.PL

0.010 Tuesday May 12 14:32:10 PDT 2009:
    - Big change: legacy-ized v005, using Path::Dispatcher to do the fun stuff
    - Can use ::Declare to do nice stuff

0.005 Thursday September 04 10:58:08 PDT 2008:
    - Properly pass through remaining arguments

0.004 Saturday August 16 20:02:09 PDT 2008:
    - Using Test::Trap for exit error code checking

0.003 Saturday August 16 15:07:24 PDT 2008: 
    - Minor documentation fix

0.002 Saturday August 16 2008:
    - Inclusion of abort functionality
    - Better documentation coverage
    - Integrated Hash::Param
    - First non-development release
    
0.001_3 Saturday August 09 10:09:44 PDT 2008:
    - Update to work with the latest Moose (reader ... instead of is ...)

0.001_2 Friday August 08 16:09:53 PDT 2008:
    - Added SOURCE information

0.001_1 Friday August 08 2008:
    - Initial release


Makefile.PL
use inc::Module::Install;

name 'Getopt-Chain';
all_from 'lib/Getopt/Chain.pm';

resources repository => 'http://github.com/robertkrimen/Getopt-Chain/tree/master';

{
    require ExtUtils::MakeMaker;
    use strict;
    no strict 'refs';

    my $libscan = \&{"ExtUtils::MM_Any::libscan"};
    *{"ExtUtils::MM_Any::libscan"} = sub {
        return '' unless $libscan->(@_);
        return '' if $_[1] =~ /\.sw[p-z]$/;
        return $_[1];
    };
}

if (-e 'inc/.author') {
    my $all_from = join '/', 'lib', split m/-/, name . '.pm';
    `perldoc -tF $all_from > README` if ! -e 'README' || (stat $all_from)[9] > (stat 'README')[9];
}

{
    map { my ($pk, $vr) = split m/\s/; build_requires $pk => $vr || 0 } grep { ! /^\s*#/ } split m/\n/, <<_END_;
Test::Most
#Test::Trap
_END_

    map { my ($pk, $vr) = split m/\s/; requires $pk => $vr || 0 } grep { ! /^\s*#/ } split m/\n/, <<_END_;
Carp::Clan::Share
Getopt::Long 2.37
Hash::Param
Moose 0.76
MooseX::AttributeHelpers
MooseX::MakeImmutable 0.02
Path::Dispatcher
_END_
}

WriteAll;


lib/Getopt/Chain.pm
package Getopt::Chain;

use warnings;
use strict;

use constant DEBUG => $ENV{GOC_TRACE} ? 1 : 0;
our $DEBUG = DEBUG;

=head1 NAME

Getopt::Chain - Command-line processing like svn and git

=head1 VERSION

Version 0.017

=cut

our $VERSION = '0.017';

=head1 SYNPOSIS 

    package My::Command;

    use Getopt::Chain::Declare;

    start [qw/ verbose|v /]; # These are "global"
                             # my-command --verbose initialize ...

    # my-command ? initialize ... --> my-command help initialize ...
    rewrite qr/^\?(.*)/ => sub { "help ".($1||'') };

    # NOTE: Rewriting applies to the command sequence, NOT options

    # my-command about ... --> my-command help about
    rewrite [ ['about', 'copying'] ] => sub { "help $1" };

    # my-command initialize --dir=...
    on initialize => [qw/ dir|d=s /], sub {
        my $context = shift;

        my $dir = $context->option( 'dir' )

        # Do initialize stuff with $dir
    };

    # my-command help
    on help => undef, sub {
        my $context = shift;

        # Do help stuff ...
        # First argument is undef because help
        # doesn't take any options
    };

    under help => sub {

        # my-command help create
        # my-command help initialize
        on [ [ qw/create initialize/ ] ] => undef, sub {
            my $context = shift;

            # Do help for create/initialize
            # Both: "help create" and "help initialize" go here
        };

        # my-command help about
        on 'about' => undef, sub {
            my $context = shift;

            # Help for about...
        };

        # my-command help copying
        on 'copying' => undef, sub {
            my $context = shift;

            # Help for copying...
        };

        # my-command help ...
        on qr/^(\S+)$/ => undef, sub {
           my $context = shift;
           my $topic = $1;

           # Catch-all for anything not fitting into the above...

           warn "I don't know about \"$topic\"\n"
        };
    };

    # ... elsewhere ...

    My::Command->new->run( [ @arguments ] )
    My::Command->new->run # Just run with @ARGV

=head1 DESCRIPTION

Getopt::Chain can be used to provide C<svn(1)>- and C<git(1)>-style option and command processing. Any option specification
covered by L<Getopt::Long> is fair game.

This is a new version of Getopt::Chain that uses L<Path::Dispatcher>

CAVEAT 1: This is pretty beta, so the sugar/interface above WILL be tweaked

CAVEAT 2: Unfortunately, Getopt::Long slurps up the entire arguments array at once. Usually, this isn't a problem (as Getopt::Chain uses 
pass_through). However, if a subcommand has an option with the same name or alias as an option for a parent, then that option won't be available
for the subcommand. For example:

    ./script --verbose --revision 36 edit --revision 48 --file xyzzy.c
    # Getopt::Chain will not associate the second --revision with "edit"

So, for now, try to use distinct option names/aliases :)

DEBUG: You can get some extra information about what Getopt::Chain is doing by setting the environment variable C<GOC_TRACE> to 1

=head1 LEGACY

The old-style, non L<Path::Dispatcher> version is still available at L<Getopt::Chain::v005>

=cut

use Moose;
use Getopt::Chain::Carp;

use Getopt::Chain::Builder;
use Getopt::Chain::Context;

has builder => qw/is ro lazy_build 1/, handles => [qw/ dispatcher /];
sub _build_builder {
    require Getopt::Chain::Builder;
    return Getopt::Chain::Builder->new;
}

has context_from => qw/is ro isa Str|CodeRef lazy_build 1/;
sub _build_context_from {
    return 'Getopt::Chain::Context';
}

sub process {
    if (! ref $_[0] && $_[0] && $_[0] eq 'Getopt::Chain') {
        shift;
        require Getopt::Chain::v005;
        carp "Deprecated: Use Getopt::Chain::v005->process( ... ) to avoid this warning (this method will be removed in a future version)";
        return Getopt::Chain::v005->process( @_ );
    }
}

sub run {
    if (! ref $_[0] ) {
        croak "Can't call run on the package; use ->new first";
    }
    my $self = shift;
    my $arguments = shift;

    $arguments = [ @ARGV ] unless $arguments;

    my $context = $self->new_context( dispatcher => $self->dispatcher, arguments => $arguments );
    $context->run;
    return $context->options;
}

sub new_context {
    my $self = shift;

    my $context_from = $self->context_from;
    if (! ref $context_from) {
        return $context_from->new( @_ );
    }
    else {
        croak "Don't understand context source \"$context_from\"";
    }
}

use MooseX::MakeImmutable;
MooseX::MakeImmutable->lock_down;

=head1 SEE ALSO

L<Getopt::Long>

L<App::Cmd>

L<MooseX::App::Cmd>

=head1 ACKNOWLEDGEMENTS

Sartak for L<Path::Dispatcher>

obra for inspiration on the CLI (via Prophet & Sd: L<http://syncwith.us/>)

=head1 AUTHOR

Robert Krimen, C<< <rkrimen at cpan.org> >>

=head1 SOURCE

You can contribute or fork this project via GitHub:

L<http://github.com/robertkrimen/Getopt-Chain/tree/master>

    git clone git://github.com/robertkrimen/Getopt-Chain.git Getopt-Chain

=head1 BUGS

Please report any bugs or feature requests to C<bug-Getopt-Chain at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Getopt-Chain>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Getopt::Chain


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Getopt-Chain>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Getopt-Chain>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Getopt-Chain>

=item * Search CPAN

L<http://search.cpan.org/dist/Getopt-Chain>

=back


=head1 COPYRIGHT & LICENSE

Copyright 2008 Robert Krimen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Getopt::Chain
